src/components/Prompt/Example.md

```diff
@@ -0,0 +1,74 @@
+`Prompt` component allows to show a dialog window if user is navigating away from a page. Usual use case for this is when user is about to leave the page with form that has changes without submitting it.
```

Comment:
I think the main example should be with a Form component. This is the most used example I suppose. I don't see such one.
Another one would be related to FileUploaderDetails.
The current solution still forces developers to write a lot of custom code to implement a prompt.

The best scenario could be using a dedicated HOC...
```tsx
// import { withPrompt } from 'precise-io';

// 1. usage with Form
const FormWithPrompt = withPrompt(Form);
<FormWithPrompt onSubmit={console.log}/>

// 2. usage with whatever else
const MyComponent = () => <div>My Component</div>
const MyComponentPrompt = withPrompt(MyComponent);
<MyComponentPrompt />
```

The main question here is how to deal with nested `withPrompt` components. For instance
```ts
const Prompt1Component = withPrompt(({children}) => <div>{{children}}</div>);
const Prompt2Component = withPrompt(() => <div>Some text</div>);
<App>
 <Prompt1Component>
  <Prompt2Component/>
 </Prompt1Component>
</App>
```

Reply:
Example can be added.

Regarding the a lot of custom code, I will have to disagree. The easiest is to just drop the `Prompt` component in your page and it will just work, as long as you have react-router context set higher above. If you don't, you can use `PromptBasic` or `PromptModal` (depending if you want browser's default prompt window or modal) and pass in text and history object. 
```jsx
// will always prompt, when you want to leave the page. 
<Prompt message="Hey, are you sure you want to leave the page" />


// this is the equivalent for 
//  const FormWithPrompt = withPrompt(Form);
// <FormWithPrompt onSubmit={console.log}/>
<Form prompt="Hey, are you sure you want to leave the page"">


// this is with precise-ui Modal
<Prompt message="Hey, are you sure you want to leave the page" modalOptions={{title: "Leaving page..."}}/>
```
For examples above, you just need to drop those components and it will work, but you also have more control (what is the text, when to show etc) in more convenient way(component rather than HOC)

And if user needs to have more customized thing, where he doesn't need modal or prompt, but button appearing on page for example, then he can just use `usePrompt` and have fully customized version.

To achieve similar things, HOC will not save a lot of custom code(maybe none), but also make it harder to control, when to enable/disable the Prompt component

regarding the question about nested components, this component has to be used with caution, since it can easily get complicated, like in scenario you wrote. 

I can give another scenario, that this component becomes problematic.
Image, you have a page, with two tabs. one tab contains list of items, second tab contains the Form. You set the form to have a prompt, so if user leaves edited form, it will ask if user really wants that. And most importantly, when user switches between tabs, url changes.

In this case, when user changes the form, and navigates to second tab, prompt will be asked, if user wants to leave, he presses ok, and navigates to second page. But, components in the first tab with form are not unmounted, the form is changed and not submitted and etc. So, when user decides to navigate again, he will be asked again, if he wants to navigate away. 

So, it is important to use the `when` prop wisely, so that prompts do not overlap, or block user's navigation to other pages

Reply:
>  then he can just use usePrompt and have fully customized version.


could you also add example on how to use `usePrompt`?

Reply:
replied as another comment below